### DISCUSSION TITLE
Please help with custom prompts!

### DISCUSSION THREAD

**Discussion by ronaldsholt** [sinaptik-ai/pandas-ai#501] [Q&A]
Hi All - I need help implementing a custom prompt. If someone could explain with a **better** example of what is in the docs it would be super helpful!

Here is what I'm doing now:
```python
# # Custom Prompts
class CustomAnalysisPrompt(Prompt):
    text = """
You are provided with a dataset that contains sales data by brand across various regions. Here's the metadata for the given pandas DataFrames:

{dataframes}

Given this data, please follow these steps:
0. Acknowledge the user's query and provide context for the analysis.
1. **Data Analysis**: < custom instructions > 
2. **Opportunity Identification**: < custom instructions > 
3. **Reasoning**: < custom instructions > 
4. **Recommendations**: < custom instructions > 
5. **Output**: Return a dictionary with:
   - type (possible values: "text", "number", "dataframe", "plot")
   - value (can be a string, a dataframe, or the path of the plot, NOT a dictionary)
   Example: {{ "type": "text", "value": < custom instructions > }}

Using the provided dataframes (`dfs`), update the Python code based on the user's query:
{conversation}

# Updated code:
# """

class CustomResponsePrompt(Prompt):
    text = """
You are provided with a dataset that contains X data. Here's the metadata for the given pandas DataFrames:

{dataframes}

your task is to summarize your findings in 3 sentences for a business executive to understand. 

response:
# """

# Creating SmartDataFrames
sdf_sales_continent = SmartDataframe(df, name="df", config={"llm": llm, 
                                                                    "non_default_prompts": 
                                                                   {"generate_python_code":CustomAnalysisPrompt(),
                                                                    "generate_response": CustomResponsePrompt()}})
```

I can't get this to work (seemingly, since it doesn't throw any errors but only returns standard responses), so I am wondering if there are any tricks here or if I am missing something. Thank you!!!

**Comment by gventuri** [comment]
Hi @ronaldsholt, thanks a lot for reporting.

I've just released a new version of the custom prompts, as the previous implementation had become too chaotic.

Here's a working example. Hope it helps.

```python
import pandas as pd
from pandasai import SmartDataframe
from pandasai.prompts import Prompt

# # Custom Prompts
class CustomAnalysisPrompt(Prompt):
    text = """
You are provided with a dataset that contains sales data by brand across various regions. Here's the metadata for the given pandas DataFrames:

{dataframes}

Given this data, please follow these steps:
0. Acknowledge the user's query and provide context for the analysis.
1. **Data Analysis**: < custom instructions > 
2. **Opportunity Identification**: < custom instructions > 
3. **Reasoning**: < custom instructions > 
4. **Recommendations**: < custom instructions > 
5. **Output**: Return a dictionary with:
   - type (possible values: "text", "number", "dataframe", "plot")
   - value (can be a string, a dataframe, or the path of the plot, NOT a dictionary)
   Example: {{ "type": "text", "value": < custom instructions > }}
``python
def analyze_data(dfs: list[pd.DataFrame]) -> dict:
   # Code goes here (do not add comments)

# Declare a result variable
result = analyze_data(dfs)
``

Using the provided dataframes (`dfs`), update the Python code based on the user's query:
{conversation}

# Updated code:
# """

df = pd.DataFrame({
    "brand": ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J"],
    "region": ["North America", "North America", "North America", "North America", "North America", "Europe", "Europe", "Europe", "Europe", "Europe"],
    "sales": [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]
})

# Creating SmartDataFrames
sdf_sales_continent = SmartDataframe(
    df,
    name="df",
    config={
        "custom_prompts": {
            "generate_python_code": CustomAnalysisPrompt(),
        }
    }
)
res = sdf_sales_continent.chat("What is the average sales by continent?")
```

If you have any questions, just ping me.

**Reply by ronaldsholt** [reply]
No longer works in 1.2.1, using AzureOpenAI. Any idea when it will be fixed?

**Reply by mspronesti** [reply]
What's not working for you?
I just tried with Gabriele's example and it does work for me.
```py
from pandasai.llm import AzureOpenAI
import pandas as pd
from pandasai import SmartDataframe
from pandasai.prompts import Prompt

# # Custom Prompts
class CustomAnalysisPrompt(Prompt):
    text = """
You are provided with a dataset that contains sales data by brand across various regions. Here's the metadata for the given pandas DataFrames:

{dataframes}

Given this data, please follow these steps:
0. Acknowledge the user's query and provide context for the analysis.
1. **Data Analysis**: < custom instructions > 
2. **Opportunity Identification**: < custom instructions > 
3. **Reasoning**: < custom instructions > 
4. **Recommendations**: < custom instructions > 
5. **Output**: Return a dictionary with:
   - type (possible values: "text", "number", "dataframe", "plot")
   - value (can be a string, a dataframe, or the path of the plot, NOT a dictionary)
   Example: {{ "type": "text", "value": < custom instructions > }}
``python
def analyze_data(dfs: list[pd.DataFrame]) -> dict:
   # Code goes here (do not add comments)

# Declare a result variable
result = analyze_data(dfs)
``

Using the provided dataframes (`dfs`), update the Python code based on the user's query:
{conversation}

# Updated code:
# """

df = pd.DataFrame({
    "brand": ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J"],
    "region": ["North America", "North America", "North America", "North America", "North America", "Europe", "Europe", "Europe", "Europe", "Europe"],
    "sales": [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]
})

llm = AzureOpenAI(
    api_version="2023-07-01-preview",
    api_token="...", # Your API token
    api_base="...", # Your API Base
    deployment_name="...", # Your model deployment name
    is_chat_model=True # this is now true by default
)

# Creating SmartDataFrames
sdf_sales_continent = SmartDataframe(
    df,
    name="df",
    config={
        "llm": llm,
        "verbose": True,
        "enable_cache": False,
        "custom_prompts": {
            "generate_python_code": CustomAnalysisPrompt(),
        }
    }
)
res = sdf_sales_continent.chat("What is the average sales by continent?")
```

**Reply by ronaldsholt** [reply]
This exact code above but using "AzureOpenAI" as the LLM - I get countless errors that I've been trying to express all on here. 

OpenAI is fine, AzureOpenAI has many problems @mspronesti

**Reply by mspronesti** [reply]
The code snippet above uses Azure OpenAI though. Unfortunately I didn't manage to reproduce your error :/

**Reply by ronaldsholt** [reply]
Hmm, thanks will try again!

**Reply by av2atgh** [reply]
On this trail... I ran the code above, using OpenAI llm. When I asked for the last prompt I get the standard pandas-ai prompt, as it ignored the custom prompt. Is it really using the custom prompt?

**Reply by adityamiskin** [reply]
@av2atgh I have a similar experience working with custom prompts. I am using a open-source langchain LLM and I cant seem to feed the custom prompt to the LLM. When can we expect a fix for this?

**Reply by vijaykeshri** [reply]
I am trying to use custom prompt because result returns always **None**. plz see log below. So thought to use custom prompt, may be that can help to get rid of None.

Any idea if this issue is fixed? It gives always default prompt text only. I am using version 1.2.7. However, custom prompts seems not working. I am also using open-source langchain LLM- Falcon 40B. Here is the log

2023-09-30 01:04:45 [INFO] Question: How many rows are there?
2023-09-30 01:04:45 [INFO] Running PandasAI with langchain_textgen LLM...
2023-09-30 01:04:45 [INFO] Prompt ID: 822fd9b7-7e92-40c6-97f6-a8789ee3
2023-09-30 01:04:45 [INFO] Using cached response
2023-09-30 01:04:45 [INFO] Code generated:
```
# TODO complete this function
def analyze_data(dfs: list[pd.DataFrame]) -> dict:
    """
    Analyze the data
    1. Prepare: Preprocessing and cleaning data if necessary
    2. Process: Manipulating data for analysis (grouping, filtering, aggregating, etc.)
    3. Analyze: Conducting the actual analysis (if the user asks to plot a chart save it to an image in exports/charts/temp_chart.png and do not show the chart.)
    4. Output: return a dictionary of:
    - type (possible values "text", "number", "dataframe", "plot")
    - value (can be a string, a dataframe or the path of the plot, NOT a dictionary)
    Example output: { "type": "text", "value": "The average loan amount is $15,000." }
    """
```

2023-09-30 01:04:45 [INFO] 
Code running:
```
def analyze_data(dfs: list[pd.DataFrame]) ->dict:
    """
    Analyze the data
    1. Prepare: Preprocessing and cleaning data if necessary
    2. Process: Manipulating data for analysis (grouping, filtering, aggregating, etc.)
    3. Analyze: Conducting the actual analysis (if the user asks to plot a chart save it to an image in exports/charts/temp_chart.png and do not show the chart.)
    4. Output: return a dictionary of:
    - type (possible values "text", "number", "dataframe", "plot")
    - value (can be a string, a dataframe or the path of the plot, NOT a dictionary)
    Example output: { "type": "text", "value": "The average loan amount is $15,000." }
    """
        ```
2023-09-30 01:04:45 [INFO] Executed in: 0.0023193359375s
None

**Reply by sivakumar41** [reply]
hi I to facing problem to use the custom prompts
can anyone help me to solve this?

**Comment by sivakumar41** [comment]
> I am trying to use custom prompt because result returns always **None**. plz see log below. So thought to use custom prompt, may be that can help to get rid of None.
> 
> Any idea if this issue is fixed? It gives always default prompt text only. I am using version 1.2.7. However, custom prompts seems not working. I am also using open-source langchain LLM- Falcon 40B. Here is the log
> 
> 2023-09-30 01:04:45 [INFO] Question: How many rows are there? 2023-09-30 01:04:45 [INFO] Running PandasAI with langchain_textgen LLM... 2023-09-30 01:04:45 [INFO] Prompt ID: 822fd9b7-7e92-40c6-97f6-a8789ee3 2023-09-30 01:04:45 [INFO] Using cached response 2023-09-30 01:04:45 [INFO] Code generated:
> 
> ```
> # TODO complete this function
> def analyze_data(dfs: list[pd.DataFrame]) -> dict:
>     """
>     Analyze the data
>     1. Prepare: Preprocessing and cleaning data if necessary
>     2. Process: Manipulating data for analysis (grouping, filtering, aggregating, etc.)
>     3. Analyze: Conducting the actual analysis (if the user asks to plot a chart save it to an image in exports/charts/temp_chart.png and do not show the chart.)
>     4. Output: return a dictionary of:
>     - type (possible values "text", "number", "dataframe", "plot")
>     - value (can be a string, a dataframe or the path of the plot, NOT a dictionary)
>     Example output: { "type": "text", "value": "The average loan amount is $15,000." }
>     """
> ```
> 
> 2023-09-30 01:04:45 [INFO] Code running:
> 
> ```
> def analyze_data(dfs: list[pd.DataFrame]) ->dict:
>     """
>     Analyze the data
>     1. Prepare: Preprocessing and cleaning data if necessary
>     2. Process: Manipulating data for analysis (grouping, filtering, aggregating, etc.)
>     3. Analyze: Conducting the actual analysis (if the user asks to plot a chart save it to an image in exports/charts/temp_chart.png and do not show the chart.)
>     4. Output: return a dictionary of:
>     - type (possible values "text", "number", "dataframe", "plot")
>     - value (can be a string, a dataframe or the path of the plot, NOT a dictionary)
>     Example output: { "type": "text", "value": "The average loan amount is $15,000." }
>     """
>         ```
> 2023-09-30 01:04:45 [INFO] Executed in: 0.0023193359375s
> None
> ```
